lib/buffer.js

```diff
@@ -7,62 +7,219 @@ const alloc = smalloc.alloc;
 const truncate = smalloc.truncate;
 const sliceOnto = smalloc.sliceOnto;
 const kMaxLength = smalloc.kMaxLength;
+const useOldBuffer = !!process.useOldBuffer;
 
-exports.Buffer = Buffer;
-exports.SlowBuffer = SlowBuffer;
-exports.INSPECT_MAX_BYTES = 50;
+var Buffer, SlowBuffer, NativeBuffer;
 
 
-Buffer.poolSize = 8 * 1024;
-var poolSize, poolOffset, allocPool;
+// New Buffer implementation using Uint8Array. Done this way so the constructor
+// name can remain "Buffer" and "SlowBuffer".
 
+if (!useOldBuffer) {
+  Buffer = function Buffer(arg) {
+    // Common case.
+    if (typeof arg === 'number') {
+      // If less than zero, or NaN.
+      if (arg < 0 || !(arg <= 0 || arg > 0))
+        arg = 0;
+      return allocate(arg);
+    }
 
-function createPool() {
-  poolSize = Buffer.poolSize;
-  allocPool = alloc({}, poolSize);
-  poolOffset = 0;
+    // Slightly less common case.
+    if (typeof arg === 'string') {
+      return fromString(arg, arguments[1]);
+    }
+
+    // Unusual.
+    return fromObject(arg);
+  };
+
+  Buffer.prototype.__proto__ = Uint8Array.prototype;
+  Buffer.__proto__ = Uint8Array;
+
+  SlowBuffer = function SlowBuffer(length) {
+    if (length < 0)
+      length = 0;
+    return binding.create(length);
+  };
+
+  SlowBuffer.prototype.__proto__ = Buffer.prototype;
+  SlowBuffer.__proto__ = Buffer;
+
+  binding.setupBufferJS(Buffer.prototype);
 }
-createPool();
 
-function Buffer(arg) {
-  if (!(this instanceof Buffer)) {
-    // Avoid going through an ArgumentsAdaptorTrampoline in the common case.
-    if (arguments.length > 1)
-      return new Buffer(arg, arguments[1]);
 
-    return new Buffer(arg);
+function allocate(size) {
+  if (size === 0)
+    return binding.create(0);
+  if (size < (Buffer.poolSize / 2)) {
```

Comment:
Did this before I realized that just because ArrayBuffers can be arbitrarily big, typed arrays can't. So it was "technically" possible to set the pool size greater than a bitwise operation could handle. But  that's not actually the case. So I'll change that back.